### ISSUE TITLE
lia is slower than omega in large contexts

### ISSUE THREAD

**Issue by andres-erbsen** [coq/coq#9848]
#### Description of the problem

```coq
Require Import Lia Omega.

Goal True.
  do 1000 pose proof I.
  Time assert (42 = 42) as _ by Lia.lia. (* 10s *)
  Time assert (42 = 42) as _ by omega. (* .01s *)
  exact I.
Abort. (* Qed is also slow here *)

Goal True.
  do 1000 pose I.
  Time assert (42 = 42) as _ by Lia.lia. (* 10s *)
  Time assert (42 = 42) as _ by omega. (* .01s *)
  exact I.
Abort. (* Qed is also slow here *)
```

Rough breakdown shows that most of the time is spent in zify, but even xlia itself is slower than omega.

```coq
Require Import Lia Omega.

Goal True.
  do 1000 pose proof I.
  assert (42 = 42).
  Time progress zify_nat. (* 1.255s *)
  Time zify_positive. (* 1.121s *)
  Time zify_N. (* 1.242s *)
  Time Fail progress zify_nat. (* 1.058s *)
  Time zify_op. (* 0s *)
  Time unfold Z.succ in *. (* 1.147s *)
  Time unfold Z.pred in *. (* 1.265s *)
  Time xlia zchecker. (* .245s *)
  exact I.
Abort. (* Qed is also slow here *)
```

Cc #7878 #7872

#### Coq Version

8.9.0, a796822e5f57f74ff36e538fd2169f70a8c6c145 (the timings are from 8.9, on master the overall timings are very similar but the fine-grained breakdown cases vary within 50%).

**Comment by samuelgruetter** [comment]
I ran into slow `lia` invocations as well and had to do manual workarounds such as `clear -H. lia.`

**Comment by samuelgruetter** [comment]
While the above example might seem artificial, there are significant performance differences between lia and omega in practice, see https://github.com/coq/coq/issues/9921

**Comment by ppedrot** [comment]
It seems `xlia` is generalizing over mere variables in a very inefficient way. I'll have a deeper look.

**Comment by fajb** [comment]
I am glad @ppedrot is optimising the generalisation code!

Yet, in the present case, the main culprit seems elsewhere: the 1000x hypotheses should simply be dropped out of the proof-term. I'll look into this...

But, is-it really representative of the slow-down? True is kind of special.... 
@samuelgruetter , do you have at hand other slow instances ?
If you could pinpoint cases over Z where omega is faster that xlia that would be much helpful.
Are there any where clear -H is needed for xlia but not for omega?

BTW, I am also working on a zify prototype #9856 -- there are still a few regressions that need  to be investigated.

**Comment by andres-erbsen** [comment]
if True is special then this is indeed off from what I intended. A high-level description of the cases where I have noticed that lia is much slower than omega or clear -linearhyps; lia is that there are many non-arithmetic hypotheses (and they are rather large). it's possible that this is property is not the source of the slowdown, I just guessed it and confirmed with the example here.

**Comment by fajb** [comment]
The lia sluggishness for the example https://github.com/coq/coq/issues/9848#issue-426221296 should be solved by commit 585f6c1 . Together with PR #9856, this is now instantaneous.
